src/System.Net.Sockets/tests/FunctionalTests/SendReceive.cs

```diff
@@ -384,6 +384,45 @@ public abstract class SendReceive<T> : SocketTestHelperBase<T> where T : SocketH
             }
         }
 
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(Loopbacks))]
+        public void SendRecv_BlockingStreamRead_Close(IPAddress listenAt)
+        {
+            // Test for #22564 when close is called while blocking read is pending.
+            using (var server = new Socket(listenAt.AddressFamily, SocketType.Stream, ProtocolType.Tcp))
+            using (var client = new Socket(listenAt.AddressFamily, SocketType.Stream, ProtocolType.Tcp))
+            {
+                server.BindToAnonymousPort(listenAt);
+                server.Listen(1);
+
+                Thread clientThread = new Thread(() =>
+                {
+                    byte[] buffer = new byte[16];
+                    client.Connect(server.LocalEndPoint);
+                    // Blocking read.
+                    try
+                    {
+                        client.Receive(buffer);
```

Comment:
It'd be good to have tests for the other main blocking operations, e.g. Send, Connect, Accept, etc.

Reply:
yes, I was planning to. There is separate bug for accept. The problem is that the shutdown does not work for that on OSX (works ok on Linux) I'm still investigating.